go/mysql/server.go

```diff
@@ -294,10 +314,28 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
 		log.Warningf("Slow connection from %s: %v", c, connectTime)
 	}
 
+	packetCh := make(chan []byte)
+	errorCh := make(chan error)
+	var data []byte
+
 	for {
+		if l.shutdown && l.handler.SafeToClose(c) {
+			return
+		}
 		c.sequence = 0
-		data, err := c.readEphemeralPacket()
-		if err != nil {
+		go func() {
+			data, err := c.readEphemeralPacket()
+			if err != nil {
+				errorCh <- err
+			} else {
+				packetCh <- data
+			}
+		}()
```

Comment:
The idea here is to make it so that `shutdownCh` can preempt `readEphemeralPacket()`. This approach feels a little gross and I'm fairly new to Go, so any pointers here are appreciated.